nav2_costmap_2d/plugins/costmap_filters/costmap_filter.cpp

```diff
@@ -0,0 +1,119 @@
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/costmap_filter.hpp"
+
+#include <exception>
+
+namespace nav2_costmap_2d
+{
+
+CostmapFilter::CostmapFilter()
+: is_rolling_(true), costmap_filter_info_topic_("/costmap_filter_info")
+{
+}
+
+CostmapFilter::~CostmapFilter()
+{
+}
+
+void CostmapFilter::onInitialize()
+{
+  // Get parameters
+  declareParameter("enabled", rclcpp::ParameterValue(true));
+  declareParameter("costmap_filter_info_topic");
+
+  node_->get_parameter(name_ + "." + "enabled", enabled_);
+  if (
+    node_->get_parameter(name_ + "." + "costmap_filter_info_topic", costmap_filter_info_topic_) ==
+    rclcpp::ParameterType::PARAMETER_NOT_SET)
+  {
+    RCLCPP_ERROR(node_->get_logger(), "costmap_filter_info_topic parameter is not set");
+    throw std::runtime_error("Parameter is not set");
+  }
+
+  // Allocate and set the costmap of current layer
+  matchSize();
+  // Ask if costmap window is rolling
+  is_rolling_ = layered_costmap_->isRolling();
+}
+
+void CostmapFilter::activate()
+{
+  initializeFilter(costmap_filter_info_topic_);
+}
+
+void CostmapFilter::deactivate()
+{
+  resetFilter();
+}
+
+void CostmapFilter::reset()
+{
+  resetMaps();
+  resetFilter();
+  initializeFilter(costmap_filter_info_topic_);
+}
+
+void CostmapFilter::updateBounds(
+  double robot_x, double robot_y, double robot_yaw,
+  double * /*min_x*/, double * /*min_y*/, double * /*max_x*/, double * /*max_y*/)
+{
+  // If window is rolling, it is required to dynamically update costmap origin of
+  // current layer
+  if (is_rolling_) {
+    updateOrigin(robot_x - getSizeInMetersX() / 2, robot_y - getSizeInMetersY() / 2);
+  }
+
+  latest_robot_x = robot_x;
+  latest_robot_y = robot_y;
+  latest_robot_yaw = robot_yaw;
+}
+
+void CostmapFilter::updateCosts(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i, int min_j, int max_i, int max_j)
+{
+  // Clear costmap_ for process()
+  resetMap(min_i, min_j, max_i, max_j);
```

Comment:
Why are you resetting the `costmap_` every iteration?

Reply:
`costmap_` could be "dirtied" on previous iterations. I've chosen to make a clean-up procedure on CostmapFilter's side in order to provide a clean `costmap_` for further usage each time when underlying `process()` API is called. This is like paint editor & canvas parallel I've mentioned above.

For optimization this could be removed, but it should be mentioned in design and header comments in order to get developers to know that the clean-up on their side. Personally, I'd like to prefer to hide all possible supplementary work (like clean-up) into CostmapFilter, but do not insist on that because of performance. What is your opinion about it?

Reply:
Control+F through the files for `costmap_->` gives me 0 hits, you don't use `costmap_` right now to even need to reset. I don't see where other costmap layers need to reset the values in its internal costmap on an update cycle (https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp), that's effectively wiping out all the data from that region which I wouldn't think you want to do. For a sensor-processing layer like obstacle / voxel, that would wipe out measurements that need to be marked and cleared by other measurements. For the static layer, I also don't see a reset of costs in the internal costmap per update when storing a static map in both rolling and non-rolling modes. 

I'm just not seeing why its needed when `costmap_` is currently unused so it can't be dirtied and isn't queried. Moreover, if/when it is used, you'd be wiping out data that you might just be updating later

Reply:
Ah, I see that the underlying layer might modify `costmap_`. Even so, I'm still not convinced that resetting it is necessary or even desirable.

Reply:
Removed in latest commit as unnecessary: https://github.com/ros-planning/navigation2/pull/1882/commits/cfeeb1a0b96560bb68fa8f8119e27a62c65e7100#diff-9abde77dd5850f2da94347098f064bf5L111